node/node_cross_link.go

```diff
@@ -58,7 +62,82 @@ func (node *Node) ProcessCrossLinkHeartbeatMessage(msgPayload []byte) {
 	if node.IsRunningBeaconChain() {
 		return
 	}
-	// process in next pr.
+
+	if err := node.processCrossLinkHeartbeatMessage(msgPayload); err != nil {
+		utils.Logger().Err(err).
+			Msg("[ProcessCrossLinkHeartbeatMessage] failed process crosslink heartbeat signal")
+	}
+}
+
+func (node *Node) processCrossLinkHeartbeatMessage(msgPayload []byte) error {
+	hb := types.CrosslinkHeartbeat{}
+	err := rlp.DecodeBytes(msgPayload, &hb)
+	if err != nil {
+		return err
+	}
+	shardID := node.Blockchain().CurrentBlock().ShardID()
+	if hb.ShardID != shardID {
+		return errors.Errorf("invalid shard id: expected %d, got %d", shardID, hb.ShardID)
+	}
+
+	// Outdated signal.
+	if hb.LatestContinuousBlockNum <= atomic.LoadUint64(&latestSentCrosslink) {
+		return nil
+	}
+
+	sig := &ffi_bls.Sign{}
+	err = sig.Deserialize(hb.Signature)
+	if err != nil {
+		return err
+	}
+
+	hb.Signature = nil
+	serialized, err := rlp.EncodeToBytes(hb)
+	if err != nil {
+		return err
+	}
+
+	pub := ffi_bls.PublicKey{}
+	err = pub.Deserialize(hb.PublicKey)
+	if err != nil {
+		return err
+	}
+
+	ok := sig.VerifyHash(&pub, serialized)
+	if !ok {
+		return errors.New("invalid signature")
+	}
+
+	state, err := node.EpochChain().ReadShardState(big.NewInt(int64(hb.Epoch)))
+	if err != nil {
+		return err
+	}
+
+	committee, err := state.FindCommitteeByID(shard.BeaconChainShardID)
+	if err != nil {
+		return err
+	}
+	pubs, err := committee.BLSPublicKeys()
+	if err != nil {
+		return err
+	}
+
+	keyExists := false
+	for _, row := range pubs {
+		if pub.IsEqual(row.Object) {
+			keyExists = true
+			break
+		}
+	}
+
+	if !keyExists {
+		return errors.New("pub key doesn't exist")
+	}
+
+	utils.Logger().Info().
+		Msgf("[ProcessCrossLinkHeartbeatMessage] storing latestSentCrosslink to %d", hb.LatestContinuousBlockNum)
+	atomic.StoreUint64(&latestSentCrosslink, hb.LatestContinuousBlockNum)
```

Comment:
latestSentCrosslink is just a temporary in-memory cache. The real lastContinuousCrossLink should be stored in DB.  Take a look at "ReadShardLastCrossLink" in the original code.

Reply:
If you don't store the latestContinuousBlockNum in DB, you should update the logic involving ReadShardLastCrossLink() since that part of the code is still reading that data from DB.